### PR TITLE
Adds a free-text filter (search) to the equipment page.

### LINKED ISSUE #8241: Equipment Page Needs Search Function
Despite the sorting options, the Equipment page still gets quite unwieldy. There should be a way to easily search for a specific piece of equipment.

### PR DESCRIPTION
**Pull request by a2lin**
Fixes #8241 

### Changes
Added a search bar to the equipment page to allow filtering of items. The search filter is implemented as a regex match on name.

Photos:
search-full
search-500px
search-cn

---
UUID: 7b8d3ea1-d46a-4e4e-8927-0dcaca995e36

### CODE REVIEW

**Comment by Alys** [comment]
I've tested this locally and it works well! It doesn't seem to interfere with any of the other functions on the page. I like it.

I'm not good enough with Angular to comment on the quality of the code or the tests needed, but other admins are. If this hasn't been commented on by any of them in about a week, I'll merge it.

Copying your question about tests to here from the issue:
"**If the solution seems acceptable i'd appreciate a pointer to what tests are needed for something like this.**"

File: website/client-old/js/controllers/inventoryCtrl.js [hunk 1]
```diff
@@ -420,3 +456,12 @@ habitrpg.controller("InventoryCtrl",
     }
   }
 ]);
+
+habitrpg.controller("InventorySearchCtrl",
+  ['$scope', 'User',
+  function($scope, User) {
+    $scope.updateEquipmentQuery = function() {
+      User.user.equipmentQuery = $scope.equipmentQuery;
```
Comment (TheHollidayInn) [inline website/client-old/js/controllers/inventoryCtrl.js]: Why are you attaching the equipment query to the user?

**Comment by TheHollidayInn** [comment]
Looks good to me. Technically, you would want to turn this into a directive and load your controller via the directive. 

For testing: Here is the InventoryCtrlSpec which has some good examples: https://github.com/HabitRPG/habitica/blob/develop/test/client-old/spec/controllers/inventoryCtrlSpec.js .

Essentially, you would want to:

1. populated 'gearByClass', 'gearByType', 'user.equipmentQuery'
2. Change the values
3. Check that $scope.filteredGearByClass and $scope.filteredGearByType have the correct values

Then, you should test that equipmentSearch will take a list of equipment and a term then return the filtered equipment.

**Comment by a2lin** [comment]
@TheHollidayInn I've fixed the issue with the User object; I made an object to hold the query in the InventoryCtrl, which I think makes more sense.

Can you explain a bit more about what you expect with regards to turning it into a directive?

Thanks for reviewing!

**Comment by TheHollidayInn** [comment]
Looking good. Here is a good article: https://teropa.info/blog/2015/10/18/refactoring-angular-apps-to-components.html#replace-ng-controller-with-component-directive

Read the section 'Replace ng-controller with Component Directive'

**Comment by a2lin** [comment]
@TheHollidayInn 
I actually seem to have discovered that no controller is needed, and have removed it. Is this better?

**Comment by TheHollidayInn** [comment]
Yes, this looks good. Let me test locally later tonight, and then I will merge.

**Comment by TheHollidayInn** [comment]
Noted towards your next Tier. Congrats on Tier 1!

**Comment by a2lin** [comment]
Thanks for the CR and tips!

**Comment by Alys** [comment]
@a2lin Finding my new Mystery Items in my huge Equipment list was SO EASY THIS MONTH!!! Thank you! :heart:

**Pull request merged**